core/node/node_framework/src/implementations/resources/pools.rs

```diff
@@ -1,175 +1,148 @@
+#[derive(Clone)]
+pub struct PoolResource<P: PoolKind> {
     connections_count: Arc<AtomicU32>,
-    builder: ConnectionPoolBuilder<Core>,
+    url: String,
+    max_connections: u32,
+    statement_timeout: Option<Duration>,
+    unbound_pool: Arc<Mutex<Option<ConnectionPool<P::DbMarker>>>>,
+    _kind: std::marker::PhantomData<P>,
 }
 
-impl Resource for MasterPoolResource {
+impl<P: PoolKind> fmt::Debug for PoolResource<P> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("PoolResource")
+            .field("connections_count", &self.connections_count)
+            .field("url", &"...")
+            .field("max_connections", &self.max_connections)
+            .field("statement_timeout", &self.statement_timeout)
+            .field("unbound_pool", &self.unbound_pool)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<P: PoolKind> Resource for PoolResource<P> {
     fn name() -> String {
-        "common/master_pool".into()
+        format!("common/{}_pool", P::kind_str())
     }
 }
 
-impl MasterPoolResource {
-    pub fn new(builder: ConnectionPoolBuilder<Core>) -> Self {
+impl<P: PoolKind> PoolResource<P> {
+    pub fn new(url: String, max_connections: u32, statement_timeout: Option<Duration>) -> Self {
         Self {
             connections_count: Arc::new(AtomicU32::new(0)),
-            builder,
+            url,
+            max_connections,
+            statement_timeout,
+            unbound_pool: Arc::new(Mutex::new(None)),
+            _kind: std::marker::PhantomData,
         }
     }
 
-    pub async fn get(&self) -> anyhow::Result<ConnectionPool<Core>> {
-        let result = self.builder.build().await;
+    fn builder(&self) -> ConnectionPoolBuilder<P::DbMarker> {
+        let mut builder = ConnectionPool::builder(&self.url, self.max_connections);
+        builder.set_statement_timeout(self.statement_timeout);
+        builder
+    }
 
-        if result.is_ok() {
-            self.connections_count
-                .fetch_add(self.builder.max_size(), Ordering::Relaxed);
-            let total_connections = self.connections_count.load(Ordering::Relaxed);
+    pub async fn get(&self) -> anyhow::Result<ConnectionPool<P::DbMarker>> {
+        let mut unbound_pool = self.unbound_pool.lock().await;
+        if let Some(pool) = unbound_pool.as_ref() {
             tracing::info!(
-                "Created a new master pool. Master pool total connections count: {total_connections}"
+                "Provided a new copy of an existing {} unbound pool",
+                P::kind_str()
             );
+            return Ok(pool.clone());
         }
-
-        result
+        let pool = self.builder().build().await?;
+        *unbound_pool = Some(pool.clone());
+
+        let old_count = self
+            .connections_count
+            .fetch_add(self.max_connections, Ordering::Relaxed);
+        let total_connections = old_count + self.max_connections;
```

Comment:
nit: does it make sense to keep `total_connections` field? If I understand correctly it's equal to `(self.unbound_pool.is_some() as u32) * self.max_connections`

Reply:
Not really. `unbound_pool`, if created, has `max_connections`, but if it's created -- it is shared among all the requesters (e.g. if two components require `unbound_pool`, we're still at `max_connections`). However, we can go beyond `max_connections` if we create custom pools (e.g. singleton ones) -- and `total_connections` is normally `max_connections + how_many_custom_connections_have_been_requested`.